Discuss lite-functions and aliases in the coding guide (#1658)
PlasmaPy/PlasmaPy: docs/contributing/coding_guide.rst

>>> d_i = (c/omega_pi).to(u.m, equivalencies=u.dimensionless_angles())    # doctest: +SKIP

.. _aliases:

Aliases
=======

> An :term:`alias` is an abbreviated version of a commonly used function
> that is intended for interactive use. For example,
> `~plasmapy.formulary.speeds.va_` is an alias to
> `~plasmapy.formulary.speeds.Alfven_speed`.

Reviewer: ```suggestion
An :term:`alias` is an abbreviated version of a commonly used function.
For example, `~plasmapy.formulary.speeds.va_` is an alias to
`~plasmapy.formulary.speeds.Alfven_speed`.  It is intended to give
users the option for shortening their code while maintaining
readability and explicit meaning.  As such, :term:`aliases` are given
to functionality that already has a widely used symbol in the
community's literature.
```

[semi-optional]  The phrase "...intended for interactive use." threw me off because it does not seem to have a real meaning.  Trying to come up with something else, sent me down this rabbit hole.

Here is a minimal example of an alias ``f_`` to ``function`` as would be
defined in :file:`plasmapy/subpackage/module.py`.

.. code-block:: python

   __all__ = ["function"]
   __aliases__ = ["f_"]

   __all__ += __aliases__

   def function():
       ...

   f_ = function
   """Alias to `~plasmapy.subpackage.module.function`."""

* Aliases should only be defined for the most commonly used functions.

* An alias should be defined immediately after the original function.

* The name of an alias should in some way indicate what the alias is
  for. For example, `~plasmapy.formulary.lengths.cwp_` is a shortcut for
  for :math:`c/ω_p`\ .

* The name of an alias should end with a trailing underscore.

* Each alias should have a one-line docstring that refers users to the
  original function.

> * The name of the main function should be included in ``__all__`` near
>   the top of each module, and the name of the alias should be included
>   in ``__aliases__``, which will then get appended to ``__all__``.

Reviewer: ```suggestion
* The name of the main function should be included in ``__all__`` near
  the top of each module, and the name of the alias should be included
  in ``__aliases__``, which will then get appended to ``__all__``.
  This is done so both the :term:`alias` and its parent function get
  properly documented.
```

.. _lite-functions:

Lite Functions
==============

Most functions in `plasmapy.formulary` accept |Quantity| instances as
arguments and use |validate_quantities| to verify that |Quantity|
arguments are valid. The use of |Quantity| operations and validations do
not noticeably impact performance during typical interactive use, but
the performance penalty can become significant for numerically intensive
applications.

> A :term:`lite-function` is a lightweight version of another `plasmapy`

Reviewer: ```suggestion
A :term:`lite-function` is a lite-weight version of another `plasmapy`
```

   is vital to double-check your implementation!

Here is a minimal example of a lite-function ``function_lite`` that
corresponds to ``function`` as would be defined in
:file:`plasmapy/subpackage/module.py`.

.. code-block:: python

   __all__ = ["function"]
   __lite_funcs__ = ["function_lite"]

   from numba import njit
   from numbers import Real
   from plasmapy.utils.decorators import bind_lite_func, preserve_signature

   __all__ += __lite_funcs__

   @preserve_signature
   @njit
   def function_lite(v: Real) -> Real:
       """
       The lite-function which accepts and returns real numbers in
       assumed SI units.
       """
       ...

   @bind_lite_func(function_lite)
   def function(v):
       """A function that accepts and returns Quantity arguments."""
       ...

* The name of each lite-function should be the name of the original
  function with ``_lite`` appended at the end. For example,
  `~plasmapy.formulary.speeds.thermal_speed_lite` is the lite-function
  associated with `~plasmapy.formulary.speeds.thermal_speed`.

* Lite-functions assume SI units for all of arguments that represent
  physical quantities.

> * Lite-functions should be defined immediately before the normal version
>   of the function.
> 
> * Lite-functions are bound to their normal version as the ``lite``
>   attribute using the
>   `~plasmapy.utils.decorators.lite_func.bind_lite_func` decorator.

Reviewer: ```suggestion
* Lite-functions should be defined immediately before the normal version
  of the function.

* Lite-functions should be used by their associate non-lite counterpart,
  except for well reasoned exceptions.  This is done to reduce code
  duplication.

* Lite-functions are bound to their normal version as the ``lite``
  attribute using the
  `~plasmapy.utils.decorators.lite_func.bind_lite_func` decorator.  This
  allows the lite function to also be accessed like
  ``thermal_speed.lite()``.
```

> * Each lite-function should be decorated with
>   `~plasmapy.utils.decorators.helpers.preserve_signature`.

Reviewer: ```suggestion
* If a lite-function is decorated with something like ``@njit``,
  then it should also be decorated with
  `~plasmapy.utils.decorators.helpers.preserve_signature`.  This
  preserves the function signature so interpreters can still
  give hints about function arguments.
```

* A lite-function should usually be decorated with ``@numba.njit`` (or

>   the like) for just-in-time compilation. If a decorator from `numba` is

Reviewer: ```suggestion
  the like) for just-in-time compilation. If a decorator from |numba| is
```

   from numbers import Real
   from plasmapy.utils.decorators import bind_lite_func, preserve_signature

   __all__ += __lite_funcs__

   @preserve_signature
   @njit
   def function_lite(v: Real) -> Real:
       """
       The lite-function which accepts and returns real numbers in
       assumed SI units.
       """
       ...

   @bind_lite_func(function_lite)
   def function(v):
       """A function that accepts and returns Quantity arguments."""
       ...

* The name of each lite-function should be the name of the original
  function with ``_lite`` appended at the end. For example,
  `~plasmapy.formulary.speeds.thermal_speed_lite` is the lite-function
  associated with `~plasmapy.formulary.speeds.thermal_speed`.

* Lite-functions assume SI units for all of arguments that represent
  physical quantities.

* Lite-functions should be defined immediately before the normal version
  of the function.

* Lite-functions are bound to their normal version as the ``lite``
  attribute using the
  `~plasmapy.utils.decorators.lite_func.bind_lite_func` decorator.

* Each lite-function should be decorated with
  `~plasmapy.utils.decorators.helpers.preserve_signature`.

> * A lite-function should usually be decorated with ``@numba.njit`` (or
>   the like) for just-in-time compilation. If a decorator from `numba` is
>   not able to be used, then an alternative is Cython_.

Reviewer: ```suggestion
* When possible, a lite-function should incorporate `numba's
  just-in-time compilation 
  <https://numba.pydata.org/numba-doc/latest/reference/jit-compilation.html>`__
  or utilize Cython_.  At a minimum any "extra" code beyond the raw
  calculation should be removed.
```

   __all__ += __lite_funcs__

   @preserve_signature
   @njit
   def function_lite(v: Real) -> Real:
       """
       The lite-function which accepts and returns real numbers in
       assumed SI units.
       """
       ...

   @bind_lite_func(function_lite)
   def function(v):
       """A function that accepts and returns Quantity arguments."""
       ...

* The name of each lite-function should be the name of the original
  function with ``_lite`` appended at the end. For example,
  `~plasmapy.formulary.speeds.thermal_speed_lite` is the lite-function
  associated with `~plasmapy.formulary.speeds.thermal_speed`.

* Lite-functions assume SI units for all of arguments that represent
  physical quantities.

* Lite-functions should be defined immediately before the normal version
  of the function.

* Lite-functions should be used by their associate non-lite counterpart,
  except for well reasoned exceptions.  This is done to reduce code
  duplication.

* Lite-functions are bound to their normal version as the ``lite``
  attribute using the
  `~plasmapy.utils.decorators.lite_func.bind_lite_func` decorator.  This

>   allows the lite function to also be accessed like

Reviewer: ```suggestion
  allows the lite-function to also be accessed like
```

  ``thermal_speed.lite()``.

* If a lite-function is decorated with something like ``@njit``,
  then it should also be decorated with
  `~plasmapy.utils.decorators.helpers.preserve_signature`.  This
  preserves the function signature so interpreters can still
  give hints about function arguments.

> * When possible, a lite-function should incorporate `numba's
>   just-in-time compilation
>   <https://numba.pydata.org/numba-doc/latest/reference/jit-compilation.html>`__
>   or utilize Cython_.  At a minimum any "extra" code beyond the raw
>   calculation should be removed.
> 

Reviewer: ```suggestion
* When possible, a lite-function should incorporate `numba's
  just-in-time compilation
  <https://numba.pydata.org/numba-doc/latest/reference/jit-compilation.html>`__
  or utilize Cython_.  At a minimum any "extra" code beyond the raw
  calculation should be removed.

* The name of the original function should be included in ``__all__``
  near the top of each module, and the name of the lite-function
  should be included in ``__lite_funcs__``, which will then get
  appended to ``__all__``. This is done so both the :term:`lite-function`
  and the original function get properly documented.
```
